Agent: Unit tests for random.rs

etrunko opened the pull request:
Expands MR #1677 with the suggestions from @jodh-intel.
Fixes #291
cc @wainersm @fidencio

jodh-intel reviewed (approved):
Thanks @etrunko !

lgtm

jodh-intel commented:
Ugh - the sizing label GHA is failing with a weird Ubuntu packaging script error:

```
2022-03-18T15:00:37.4370855Z Processing triggers for man-db (2.9.1-1) ...
2022-03-18T15:00:43.2334743Z ##[error]Process completed with exit code 1.
```

fidencio commented:
/test

etrunko commented:
/test

wainersm commented on src/agent/src/random.rs:
```diff
@@ -13,7 +13,7 @@ use tracing::instrument;
 pub const RNGDEV: &str = "/dev/random";
 pub const RNDADDTOENTCNT: libc::c_int = 0x40045201;
-pub const RNDRESEEDRNG: libc::c_int = 0x5207;
+pub const RNDRESEEDCRNG: libc::c_int = 0x5207;
```
Not an expert in kernel nor security writing...
... is CRNG (Cryptographic RNG) same as RNG? Does `/dev/random` device uses a cryptographic algorithm?

jodh-intel replied:
@wainersm - this is correcting a typo - see:

- https://github.com/torvalds/linux/blob/master/include/uapi/linux/random.h#L39

In fact, it might be a good idea @etrunko if you added a link to the linux source file in a comment to make this clearer.

wainersm commented on src/agent/src/random.rs:
```diff
@@ -46,3 +46,24 @@ pub fn reseed_rng(data: &[u8]) -> Result<()> {
     Ok(())
 }
+
+#[cfg(test)]
```
I wrote this months ago and I already forgot it. I assume it is okay :D

wainersm commented on src/agent/src/random.rs:
```diff
@@ -66,4 +66,20 @@ mod tests {
+        // Ensure the buffer was filled.
+        assert!(n == POOL_SIZE);
+        let ret = reseed_rng(&seed);
+        if nix::unistd::Uid::effective().is_root() {
```
The other `test_reseed_rng()` will run if it is root. So maybe this one you should just skip if root.

jodh-intel replied:
Yep - @etrunko it would be better to remove this `if` test and instead assert that this test is not running as root at the start of the function.

wainersm commented on src/agent/src/random.rs:
```diff
@@ -13,7 +13,7 @@ use tracing::instrument;
 
 pub const RNGDEV: &str = "/dev/random";
```
I'm wondering if we really want to use `/dev/random` as it blocks the caller (as opposed to `urandom`). See https://man7.org/linux/man-pages/man4/random.4.html

Hi @bergwolf , have you implemented this first? Any comment?
Cc @jodh-intel

jodh-intel replied:
@wainersm - possibly, but more likely `getrandom(2)`. Certainly if there is insufficient entropy to boot the guest kernel we probably shouldn't continue anyway. However, all that is out of scope of this PR, which is primarily for a test change.

wainersm commented on src/agent/src/random.rs:
```diff
@@ -66,4 +66,20 @@ mod tests {
+        if nix::unistd::Uid::effective().is_root() {
+            assert!(ret.is_ok());
+        } else {
+            assert!(!ret.is_ok());
```
You could have used `is_err` instead? Hey, I did my homework of Rust before this review :D

jodh-intel replied:
Yep - either wfm.

wainersm commented:
@etrunko left a few comments, mostly nits. Your first contribution to this repo, yay!

jodh-intel commented:
@etrunko - could you add that kernel link so we can at last land this? :)

etrunko commented:
> @etrunko - could you add that kernel link so we can at last land this? :)

Sorry, I got sidetracked this week, but I'll be addressing the comments asap.

fidencio commented:
@etrunko, any news on this one?

liubin commented:
/test

arronwy commented:
Can one of the admins verify this patch?

jodh-intel commented:
/retest

jodh-intel commented:
The only failing CI is ARM, due to #4150. That's unrelated to this PR so I'm going to force this PR...

/retest-arm

/cc @jongwu.

jodh-intel commented:
Greeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeennnn! ;)
